### PR TITLE
Add reference for ProviderConfig objects and deprecate Provider reference

### PR DESCRIPTION
**Pull request by muvaf**
### Description of your changes

We have decided to use `ProviderConfig` for the object that contains the information about how the provider should authenticate itself to its API.

Part of https://github.com/crossplane/crossplane/issues/1684

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [x] Updated any relevant [documentation], [examples], or [release notes].
- [x] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplane/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplane/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

File: pkg/resource/resource.go [hunk 1]
```diff
@@ -360,7 +360,10 @@ func GetExternalTags(mg Managed) map[string]string {
 	if mg.GetClassReference() != nil {
 		tags[ExternalResourceTagKeyClass] = mg.GetClassReference().Name
 	}
-	if mg.GetProviderReference().Name != "" {
+	switch {
+	case mg.GetProviderConfigReference() != nil && mg.GetProviderConfigReference().Name != "":
+		tags[ExternalResourceTagKeyProvider] = mg.GetProviderConfigReference().Name
+	case mg.GetProviderReference() != nil && mg.GetProviderReference().Name != "":
```
Comment (hasheddan) [inline pkg/resource/resource.go:366]: nitpick: would you mind adding a note or todo here to remove this branch once `Provider` type is removed?

File: apis/core/v1alpha1/resource.go [hunk 2]
```diff
@@ -197,9 +197,14 @@ type ResourceSpec struct {
 	// +optional
 	ClassReference *corev1.ObjectReference `json:"classRef,omitempty"`
 
+	// ProviderConfigReference specifies the provider that will be used to create,
+	// observe, update, and delete this managed resource.
```
Comment (negz) [inline apis/core/v1alpha1/resource.go:201]:
```suggestion
	// ProviderConfigReference specifies how the provider that will be used to create,
	// observe, update, and delete this managed resource should be configured.
```

File: apis/core/v1alpha1/resource.go [hunk 3]
```diff
@@ -197,9 +197,14 @@ type ResourceSpec struct {
...
+	ProviderConfigReference *Reference `json:"providerConfigRef,omitempty"`
```
Comment (negz) [inline apis/core/v1alpha1/resource.go:202]: Do you think we could get away with something shorter like `pcRef` here? `providerConfigRef` is probably better - just thinking out loud.

Comment (muvaf) [reply apis/core/v1alpha1/resource.go:202]: `pcRef` sounds a bit cryptic to me. I'd go with `providerConfigRef`.

**Comment by muvaf** [comment]
@hasheddan @negz all comments are addressed. Could you look at it again?

**Pull request merged**